packages/framework/src/Model/Product/Search/ProductElasticsearchRepository.php

```diff
@@ -47,26 +47,32 @@ class ProductElasticsearchRepository
      * @param \Elasticsearch\Client $client
      * @param \Shopsys\FrameworkBundle\Model\Product\Search\ProductElasticsearchConverter $productElasticsearchConverter
      * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\ElasticsearchStructureManager $elasticsearchStructureManager
+     * @param \Shopsys\FrameworkBundle\Model\Product\Search\FilterQueryFactory|null $filterQueryFactory
      */
     public function __construct(
         string $indexPrefix,
         Client $client,
         ProductElasticsearchConverter $productElasticsearchConverter,
-        ElasticsearchStructureManager $elasticsearchStructureManager
+        ElasticsearchStructureManager $elasticsearchStructureManager,
+        ?FilterQueryFactory $filterQueryFactory = null
     ) {
         $this->indexPrefix = $indexPrefix;
         $this->client = $client;
         $this->productElasticsearchConverter = $productElasticsearchConverter;
         $this->elasticsearchStructureManager = $elasticsearchStructureManager;
-        $this->filterQueryFactory = $this->createFilterQueryFactory();
+        $this->filterQueryFactory = $filterQueryFactory ?? $this->createFilterQueryFactory();
+
     }
 
     /**
+     * @required
      * @return \Shopsys\FrameworkBundle\Model\Product\Search\FilterQueryFactory
      * @deprecated Will be replaced with constructor injection in the next major release
      */
     protected function createFilterQueryFactory(): FilterQueryFactory
```

Comment:
Method has to be public to use `@required` annotation, but according to https://symfony.com/doc/current/contributing/code/bc.html#changing-classes this type of change is a violation of BC promise. 

But if you call it directly in constructor, you don't need `@required` at all, right?

Reply:
You are absolutely right, I forgot it there